gocp/gocp.go

```diff
@@ -0,0 +1,84 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"os"
+	"strings"
+)
+
+// copy copies a Go file from src to dst
+// with package renamed to pkg.
+func copy(dest, pkg, src string) error {
+	if dest == "" {
+		return errors.New("gocp: destination filepath must be set")
+	}
+	if pkg == "" {
+		return errors.New("gocp: destination package name must be set")
+	}
+	if src == "" {
+		return errors.New("gocp: source filepath must be set")
+	}
+
+	inFile, err := os.Open(src) //nolint:gosec // This is a file generation tool that takes filepaths as input and output.
+	if err != nil {
+		return err
+	}
+	defer inFile.Close()
+
+	outFile, err := os.Create(dest) //nolint:gosec // This is a file generation tool that takes filepaths as input and output.
+	if err != nil {
+		return err
+	}
+	defer outFile.Close()
+
+	w := bufio.NewWriter(outFile)
+	if _, err := w.WriteString("// Code generated by gocp. DO NOT EDIT.\n\n"); err != nil {
+		return err
+	}
+	scanner := bufio.NewScanner(inFile)
+	pkgReplaced := false // Rename the line with package statement only once.
+	for scanner.Scan() {
+		txt := scanner.Text()
+
+		if !pkgReplaced && strings.HasPrefix(txt, "package ") {
```

Comment:
This seems like it has limited functionality compared to using a text template. When these copied internal packages start to depend on each other the imports will need to be appropriately templatized. Why not just use a text template for this?

Reply:
> When these copied internal packages start to depend on each other the imports will need to be appropriately templatized.

That is true. Currently, we already have a use case for it: https://github.com/open-telemetry/opentelemetry-go/blob/4ccd59056982f34c3fcc8faef574f209b70d0f59/sdk/internal/env/env_test.go#L24

>  Why not just use a text template for this?

I found it easier to use actual Go code than writing Go source code templates because we get IDE support. 

The second reason is that the the "shared Go code" is analyzed using `golangci-lint`  what we could not say about templates. The problem is that if the generated code has `// Code generated by gocp. DO NOT EDIT.` comment then [it won't be analyzed by `golangci-lint`](https://github.com/golangci/golangci-lint/blob/076f6b911d718cf9845fe338d40972f41510c4fe/pkg/result/processors/autogenerated_exclude.go#L72-L93). We could alter the text to something like `// Generated by gocp.` However, we will lint all the generated files instead of only linting the source files .

**Summary**

Having all that said I feel that we indeed need this flexibility and I will try to simply create a simple CLI for text/template.